plutus-core/plutus-ir/src/PlutusIR/Compiler/Datatype.hs

```diff
@@ -121,78 +122,153 @@ We also need to think about the types. In general, we need:
-    (t_c_1_1 -> .. -> t_c_1_(c_1_k) -> out_T) -> .. -> (t_c_j_1 -> .. -> t_c_j_(c_j_k) -> out_T) ->
+    (t_c_0_0 -> .. -> t_c_0_(c_0_k) -> out_T) -> .. -> (t_c_j_0 -> .. -> t_c_j_(c_j_k) -> out_T) ->
       out_T
 
 The type of the constructor c_i is:
-forall (t_1 :: *) .. (t_n :: *) .
-  t_c_i_1 -> .. -> t_c_i_(c_i_k) ->
-    (T t_1 .. t_n)
+forall (t_0 :: *) .. (t_n :: *) .
+  t_c_i_0 -> .. -> t_c_i_(c_i_k) ->
+    (T t_0 .. t_n)
 This is actually declared for us in the datatype declaration, but without the type
 variables being abstracted out. We use this to get the argument types of the constructor.
 
 The term for the constructor c_i is:
-/\(t_1 :: *) .. (t_n :: *) .
-  \(a_1 : t_c_j_1) .. (a_c_i_(c_i_k) : t_c_i_(c_i_k)) .
-    abs out_T :: * .
-      \(case_c_1 : (t_c_1_1 -> .. -> t_c_1_(c_1_k) -> out_T)) .. (case_c_j : (t_c_j_1 -> .. -> t_c_j_(c_j_k) -> out_T)) .
-        case_c_i a_1 .. a_c_i_(c_i_k)
+/\(t_0 :: *) .. (t_n :: *) .
+  \(a_0 : t_c_j_0) .. (a_c_i_(c_i_k) : t_c_i_(c_i_k)) .
+    /\(out_T :: *) .
+      \(case_c_0 : (t_c_0_0 -> .. -> t_c_0_(c_1_k) -> out_T)) .. (case_c_j : (t_c_j_0 -> .. -> t_c_j_(c_j_k) -> out_T)) .
+        case_c_i a_0 .. a_c_i_(c_i_k)
 
 The type of the destructor is:
-forall (t_1 :: *) .. (t_n :: *) .
-  (T t_1 .. t_n) ->
+forall (t_0 :: *) .. (t_n :: *) .
+  (T t_0 .. t_n) ->
     forall out_T :: * .
-      (t_c_1_1 -> .. -> t_c_1_(c_1_k) -> out_T) -> .. -> (t_c_j_1 -> .. -> t_c_j_(c_j_k) -> out_T) ->
+      (t_c_0_0 -> .. -> t_c_0_(c_0_k) -> out_T) -> .. -> (t_c_j_0 -> .. -> t_c_j_(c_j_k) -> out_T) ->
         out_T
 
 The term for the destructor is:
-/\(t_1 :: *) .. (t_n :: *) .
-  \(x :: (T t_1 .. t_n)) ->
+/\(t_0 :: *) .. (t_n :: *) .
+  \(x :: (T t_0 .. t_n)) .
     x
+-}
+
+{- Note [SOP encoding of datatypes]
+We can translate our datatypes using SOPs. This is much more direct than Scott encoding, since
+we have terms in the language for construction and destruction themselves.
+
+We still need to think about the types. In general, we need:
+- The encoded type corresponding to the datatype itself
+- The encoded type of each constructor
+- The encoded term for each constructor
+- The encoded type of the destructor
+- The encoded term for the destructor
+
+-- Basic example
+
+For example, consider 'data Maybe a = Nothing | Just a'. Then:
+- The type corresponding to the datatype is:
+  Maybe = \(a :: *) -> sop [] [a]
+- The type of the constructors are:
+  Just : forall (a :: *) . a -> Maybe a
+  Nothing : forall (a :: *) . Maybe a
+- The terms for the constructors are:
+  Just = \(arg1 : a). constr (Maybe a) 1 a
+  Nothing = constr 0 (Maybe a)
+- The type of the destructor is:
+  match_Maybe : forall (a :: *) . Maybe a -> forall (out_Maybe :: *) . out_Maybe -> (a -> out_Maybe) -> out_Maybe
+- The term for the destructor is:
+  match_Maybe = /\(a :: *) -> \(x : Maybe a) -> /\(out_Maybe :: *) -> \(case_Nothing :: out_Maybe) (case_Just :: a -> out_Maybe) -> case out_Maybe x case_Nothing case_Just
```

Comment:
Is that really the term and not `(\(x' : a)  -> case_Just x'` instead of just `case_Just`?

Reply:
Yes, it doesn't have to be eta-expanded